lib/ansible/modules/network/fortimanager/fmgr_ha.py

```diff
@@ -288,37 +250,40 @@ def main():
         "fmgr_ha_file_quota": module.params["fmgr_ha_file_quota"],
         "fmgr_ha_cluster_id": module.params["fmgr_ha_cluster_id"],
     }
+    module.paramgram = paramgram
+    fmgr = None
+    if module._socket_path:
+        connection = Connection(module._socket_path)
+        fmgr = FortiManagerHandler(connection, module)
+        fmgr.tools = FMGRCommon()
+    else:
+        module.fail_json(**FAIL_SOCKET_MSG)
+
     # INIT FLAGS AND COUNTERS
     get_ha_peers = 0
     # validate required arguments are passed; not used in argument_spec to allow params to be called from provider
-    # check if params are set
-    if module.params["host"] is None or module.params["username"] is None:
-        module.fail_json(msg="Host and username are required for connection")
-    # CHECK IF LOGIN FAILED
-    fmg = AnsibleFortiManager(module, module.params["host"], module.params["username"], module.params["password"],)
-    response = fmg.login()
-    if response[1]['status']['code'] != 0:
-        module.fail_json(msg="Connection to FortiManager Failed")
-    else:
-        # START SESSION LOGIC
+
+    results = DEFAULT_RESULT_OBJ
+    try:
+        # RULE VALIDATIONS!
         # IF THE PEER SN DEFINED, BUT THE IPS ARE NOT, THEN QUIT
         if paramgram["fmgr_ha_peer_sn"] is not None:
             # CHANGE GET_HA_PEERS TO SHOW INTENT TO EDIT PEERS
             get_ha_peers = 1
             # DOUBLE CHECK THAT THE REST OF THE NEEDED PARAMETERS ARE THERE
             if paramgram["fmgr_ha_peer_ipv4"] is None and paramgram["fmgr_ha_peer_ipv6"] is None:
-                fmgr_logout(fmg, module, msg="HA Peer Serial Number is defined but the "
-                                             "IPv4 and IPv6 fields are empty."
-                                             " Fill in the IPv4 or v6 parameters in the playbook")
+                module.exit_json(msg="HA Peer Serial Number is defined but the "
```

Comment:
Can `required_together` be used to handle these cases instead of `exit_json` calls?

Reply:
Ended up using required_if, but yes we were able to remove about 35 lines worth of validation code in favor of required_if.

Reply:
Great!